ffcx/codegeneration/ufc.h

```diff
@@ -382,52 +396,66 @@ extern "C"
-  ///         cell geometries for both cells sharing the facet must be provided.
-  /// @param[in] entity_local_index
-  ///         Local index of mesh entity on which to tabulate.
-  ///         This applies to facet integrals.
-  /// @param[in] cell_orientation
-  ///         Sign of orientation of the cell with respect to the consistent
-  ///         orientation of the whole mesh.
-  ///         0 means "up"
-  ///         1 means "down" and scales det(J) with -1.0
-  ///         Applies to the case of k-dimensional surface in n-dimensional space, where k < n.
+  ///         Restriction dimension applies to interior facet integrals,
+  ///         where cell geometries for both cells sharing the facet
+  ///         must be provided.
+  /// @param[in] entity_local_index Local index of mesh entity on which
+  ///         to tabulate. This applies to facet integrals.
+  /// @param[in] quadrature_permutation For facet integrals, numbers to
+  ///         indicate the permutation to be applied to each side of the
+  ///         facet to make the orientations of the faces matched up
+  ///         should be passed in. If an integer of value N is passed
+  ///         in, then:
+  ///
+  ///          - floor(N / 2) gives the number of rotations to apply to the
+  ///          facet
+  ///          - N % 2 gives the number of reflections to apply to the facet
   ///
-  typedef void (ufc_tabulate_tensor)(ufc_scalar_t* restrict A,
-      const ufc_scalar_t* w,
-      const ufc_scalar_t* c,
-      const double* restrict coordinate_dofs,
-      const int* entity_local_index,
-      const int* cell_orientation);
+  ///         For integrals not on facets, this argument has not effect
+  ///         and a null pointer can be passed. For
+  ///         interior facets the array will have size 2 (one permutation
+  ///         for each cell adjacent to the facet). For exterior facets,
+  ///         this will have size 1.
+  /// @param[in] edge_reflections An array of bools to say whether or not each
+  ///         each edge has been reflected. This is used to ensure that vector
+  ///         dofs are correctly oriented. This array will have one entry for
+  ///         each edge of the cell.
+  /// @param[in] face_reflections An array of bools to say whether or not each
+  ///         each face has been reflected. This is used to ensure that vector
+  ///         dofs are correctly oriented. This array will have one entry for
+  ///         each face of the cell.
+  /// TODO: Use std::vector<int32_t> to store 1/0 marker for each edge/face
+  typedef void(ufc_tabulate_tensor)(
+      ufc_scalar_t* restrict A, const ufc_scalar_t* w, const ufc_scalar_t* c,
+      const double* restrict coordinate_dofs, const int* entity_local_index,
+      const uint8_t* restrict quadrature_permutation,
+      const bool* edge_reflections, const bool* face_reflections,
+      const uint8_t* face_rotations);
```

Comment:
You didn't document `face_rotations`. Also, it feels a bit weird now. If I understand correctly, in general, for vector dofs you need to know `face_rotations` for a cell integrals too, right? So for a interior facet integrals, isnt now information about rotations of codim 1 entities, facets, duplicated?

Reply:
`quadrature_permutation` has information about the rotation from two cells either side, while the other inputs are information about the different entities on one cell. But I agree that restructuring what these inputs are at some point is a good idea.